compiler/rustc_passes/src/reachable.rs

```diff
@@ -1,9 +1,13 @@
-// Finds items that are externally reachable, to determine which items
-// need to have their metadata (and possibly their AST) serialized.
-// All items that can be referred to through an exported name are
-// reachable, and when a reachable thing is inline or generic, it
-// makes all other generics or inline functions that it references
-// reachable as well.
+//! Finds local items that are externally reachable, to determine which items
+//! need to have their metadata (and possibly their AST) serialized.
+//!
+//! An item is "externally reachable" if it is relevant for other crates. This obviously includes
+//! all public items. However, when a reachable function `f` of this crate has its code generated as
+//! part of another crate's compilation, that other crate must be able to access everything needed
+//! to compile `f` -- in particular, every function called by `f`! Therefore, when a reachable thing
+//! is cross-crate inlinable or generic, it makes all other functions that it references reachable
+//! as well. Similarly, `const fn` may be evaluated by other crates, so everything they call is also
+//! "externally reachable". This is how private items can become "externally reachable".
```

Comment:
"evaluated" here refers to "executed in the compile-time interpreter". That's only possible if the MIR for these functions is available.

Reply:
But then why does it include `const fn` in "needs cross-crate MIR"?

Reply:
Turns out `const fn` are included because they can return function pointers and that can make otherwise unreachable functions reachable. I have updated the comments.